File: Anul II/Medii de proiectare si programare/Laborator/lab06/services/src/main/java/com/ubb/service/RaceService.java
package com.ubb.service;

import com.ubb.model.Race;
import com.ubb.model.RaceEntry;
import com.ubb.model.data.RaceDTO;
import com.ubb.repository.IRaceEntryRepository;
import com.ubb.repository.IRaceRepository;
import com.ubb.validation.IValidator;
import com.ubb.validation.RaceValidator;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;

public class RaceService implements IRaceService {
    private final IRaceRepository raceRepository;
    private final IRaceEntryRepository raceEntryRepository;
    private final IValidator<Race> validator;

    public RaceService(final IRaceRepository raceRepository, final IRaceEntryRepository raceEntryRepository) {
        this.raceEntryRepository = raceEntryRepository;
        this.raceRepository = raceRepository;
        validator = new RaceValidator();
    }

    @Override
    public Collection<RaceDTO> getRacesWithParticipantCount() {
        Collection<RaceDTO> result = new ArrayList<>();
        Collection<Race> races = raceRepository.findAll();
        for (Race race : races) {
            result.add(new RaceDTO(race.getName(), race.getEngineCapacity(),
                    raceEntryRepository.getEntriesByRace(race.getID()).size()));
        }
        return result;
    }

    @Override
    public void saveRaceEntry(RaceEntry newEntity) {
        raceEntryRepository.save(newEntity);
    }

    @Override
    public Collection<Race> getRacesWhereNotRegisteredAndEngineCapacity(Long participantID, Integer engineCapacity) {
        return raceEntryRepository.getRacesWhereNotRegisteredAndEngineCapacity(participantID, engineCapacity);
    }

    @Override
    public Optional<Race> getRaceByName(String name) {
        return raceRepository.getRaceByName(name);
    }

    @Override
    public Collection<RaceEntry> getEntriesByRace(Long raceID) {
        return raceEntryRepository.getEntriesByRace(raceID);
    }

    @Override
    public void deleteByIds(Long participantID, Long raceID) {
        raceEntryRepository.deleteByIds(participantID, raceID);
    }

    @Override
    public Collection<Race> getRacesByEngineCapacity(Integer engineCapacity) {
        return raceRepository.getRacesByEngineCapacity(engineCapacity);
    }

    @Override
    public Optional<Race> findById(Long id) {
        return raceRepository.findById(id);
    }

    @Override
    public Collection<Race> findAll() {
        return raceRepository.findAll();
    }

    @Override
    public void save(Race newEntity) {
        validator.validateEntity(newEntity);
        raceRepository.save(newEntity);
    }

    @Override
    public void delete(Long id) {
        raceRepository.delete(id);
    }

    @Override
    public void update(Race updatedEntity) {
        validator.validateEntity(updatedEntity);
        raceRepository.update(updatedEntity);
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab07/Ubb.BikeContest.Service/src/Service/RaceService.cs
using System.Collections;
using Ubb.BikeContest.Model;
using Ubb.BikeContest.Model.data;
using Ubb.BikeContest.Repository;
using Ubb.BikeContest.Services.validation;

namespace Ubb.BikeContest.Services.src.Service;

public class RaceService : IRaceService
{
    private readonly IRaceRepository _raceRepository;
    private readonly IRaceEntryRepository _raceEntryRepository;
    private readonly IValidator<long, Race> _validator;

    public RaceService(IRaceRepository raceRepository, IRaceEntryRepository raceEntryRepository)
    {
        _raceRepository = raceRepository;
        _raceEntryRepository = raceEntryRepository;
        _validator = new RaceValidator();
    }

    public Race FindById(long id)
    {
        return _raceRepository.FindById(id);
    }

    public IEnumerable<Race> FindAll()
    {
        return _raceRepository.FindAll();
    }

    public void Save(Race newEntity)
    {
        _validator.ValidateEntity(newEntity);
        _raceRepository.Save(newEntity);
    }

    public void Delete(long id)
    {
        _raceRepository.Delete(id);
    }

    public void Update(Race updatedEntity)
    {
        _validator.ValidateEntity(updatedEntity);
        _raceRepository.Update(updatedEntity);
    }

    public Race GetRaceByName(string name)
    {
        return _raceRepository.GetRaceByName(name);
    }

    public IEnumerable<RaceEntry> GetEntriesByRace(long raceId)
    {
        return _raceEntryRepository.GetEntriesByRace(raceId);
    }

    public void DeleteByIds(long participantId, long raceId)
    {
        throw new NotImplementedException();
    }

    public IEnumerable<Race> GetRacesByEngineCapacity(int engineCapacity)
    {
        return _raceRepository.GetRacesByEngineCapacity(engineCapacity);
    }

    public IEnumerable<RaceDto> GetRacesWithParticipantCount()
    {
        IEnumerable<Race> races = _raceRepository.FindAll();
        IList<RaceDto> result = new List<RaceDto>();
        foreach (var race in races)
        {
            result.Add(new RaceDto(race.Name, race.EngineCapacity,
                _raceEntryRepository.GetEntriesByRace(race.Id).Count()));
        }
        return result;
    }

    public void SaveRaceEntry(RaceEntry newEntity)
    {
        _raceEntryRepository.Save(newEntity);
    }

    public IEnumerable<Race> GetRacesWhereNotRegisteredAndEngineCapacity(long participantId, int engineCapacity)
    {
        return _raceEntryRepository.GetRacesWhereNotRegisteredAndEngineCapacity(participantId, engineCapacity);
    }
}